File: src/bumperbot_py_examples/bumperbot_py_examples/simple_turtlesim_kinematics.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
import math

class SimpleTurtlesimKinematics(Node):
    def __init__(self):
        super().__init__('simple_turtlesim_kinematics')

        self.turtle1_pose_sub_ = self.create_subscription(Pose, "/turtle1/pose", self.turtle1_pose_callback, 10)
        self.turtle2_pose_sub_ = self.create_subscription(Pose, "/turtle2/pose", self.turtle2_pose_callback, 10)

        self.last_turtle1_pose_ = Pose()
        self.last_turtle2_pose_ = Pose()

    def turtle1_pose_callback(self, msg):
        self.get_logger().info(f"Turtle1 Pose - x: {msg.x}, y: {msg.y}, theta: {msg.theta}")
        self.last_turtle1_pose_ = msg
    def turtle2_pose_callback(self, msg):
        self.get_logger().info(f"Turtle2 Pose - x: {msg.x}, y: {msg.y}, theta: {msg.theta}")
        self.last_turtle2_pose_ = msg
        Tx = self.last_turtle2_pose_.x - self.last_turtle1_pose_.x
        Ty = self.last_turtle2_pose_.y - self.last_turtle1_pose_.y
        distance = (Tx**2 + Ty**2)**0.5
        theta_rad = self.last_turtle2_pose_.theta - self.last_turtle1_pose_.theta
        theta_deg = 180.0 * theta_rad / 3.141592653589793
        self.get_logger().info("""\n
        Translation Vector from Turtle1 to Turtle2: \n
        Tx: {:.5f}, Ty: {:.5f} \n
        Distance: {:.5f} \n
        Rotation Matrix 
        Theta radians: {:.5f},
        theta deg: {:.5f}
        |R11    R12| :|{:.5f}   {:.5f}| :|cos(theta)  -sin(theta)|
        |R21    R22| :|{:.5f}   {:.5f}| :|sin(theta)   cos(theta)|
        """.format(Tx, Ty, distance, theta_rad, theta_deg,
                    math.cos(theta_rad), -math.sin(theta_rad),
                    math.sin(theta_rad), math.cos(theta_rad)))


def main(args=None):
    rclpy.init(args=args)
    simple_turtlesim_kinematics = SimpleTurtlesimKinematics()
    try:
        rclpy.spin(simple_turtlesim_kinematics)
    except KeyboardInterrupt:
        pass
    simple_turtlesim_kinematics.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_turtlesim_kinematics.cpp
#include "bumperbot_cpp_examples/simple_turtlesim_kinematics.hpp"

using std::placeholders::_1; // one variable as input

SimpleTurtlesimKinematics::SimpleTurtlesimKinematics(const std::string &name ): Node(name)
{
    turtle1_pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
        "/turtle1/pose", 10,
        std::bind(&SimpleTurtlesimKinematics::turtle1PoseCallback, this, _1)
    );

    turtle2_pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
        "/turtle2/pose", 10,
        std::bind(&SimpleTurtlesimKinematics::turtle2PoseCallback, this, _1)
    );

}


void SimpleTurtlesimKinematics::turtle1PoseCallback(const turtlesim::msg::Pose &pose)
{
    last_turtle1_pose_ = pose;
    //RCLCPP_INFO(this->get_logger(), "Turtle1 Pose - x: %.2f, y: %.2f, theta: %.2f", 
    //            pose.x, pose.y, pose.theta);
}

void SimpleTurtlesimKinematics::turtle2PoseCallback(const turtlesim::msg::Pose &pose)
{
    last_turtle2_pose_ = pose;
    //RCLCPP_INFO(this->get_logger(), "Turtle2 Pose - x: %.2f, y: %.2f, theta: %.2f", 
    //            pose.x, pose.y, pose.theta);
    float Tx, Ty, distance;
    Tx = last_turtle2_pose_.x - last_turtle1_pose_.x;
    Ty = last_turtle2_pose_.y - last_turtle1_pose_.y;
    distance = sqrt(Tx*Tx + Ty*Ty);
    float theta_rad = last_turtle2_pose_.theta - last_turtle1_pose_.theta;
    float theta_deg = 180.0 * theta_rad / 3.141592653589793;
    RCLCPP_INFO_STREAM(this->get_logger(), 
        "Relative Orientation (Turtle2 wrt Turtle1): " << theta_deg << " degrees\n"
        << "Rotation Matrix: |R11    R12| :|" << std::cos(theta_rad) << "   " << -std::sin(theta_rad) << "| :|cos(theta)  -sin(theta)|\n"
        << "                 |R21    R22| :|" << std::sin(theta_rad) << "   " << std::cos(theta_rad)  << "| :|sin(theta)   cos(theta)|\n"
        << "Translation vector Tx: " << Tx << ", Ty: " << Ty << "\n"
        << "Distance between Turtle1 and Turtle2: " << distance);

}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto simple_turtlesim_kinematics = std::make_shared<SimpleTurtlesimKinematics>("simple_turtlesim_kinematics");
    rclcpp::spin(simple_turtlesim_kinematics);
    rclcpp::shutdown();
    return 0;
}
